### PR TITLE
Add multi operations

### PR DESCRIPTION
**Pull request by ThunderDeliverer**
I added the changes that we discussed. The only things that I'm still on the fence are:

- Should we add a multi-getter for messages to be signed as well? I think that we might want to do that to provide full functionality, but at the same time, the single getter is a view function that doesn't really cause any gas waste if called multiple times 🤔 
- Should we require that the user using presigned emote provide the signer of the message? If we don't, an incorrect address could be recovered and the signature would still pass. I see the possibility for abuse here as you could simply generate new accounts and sign emotes for them and then just send those signatures and farm emotes..

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,14 +299,26 @@ The impressions could have been done using user-supplied strings or numeric valu
 3. **Should the proposal establish an emotable extension or a common-good repository?**\
 Initially we set out to create an emotable extension to be used with any ERC-721 compilant tokens. However, we realized that the proposal would be more useful if it was a common-good repository of emotable tokens. This way, the tokens that can be reacted to are not only the new ones but also the old ones that have been around since before the proposal.\
 In line with this decision, we decided to calculate a deterministic address for the repository smart contract. This way, the repository can be used by any NFT collection without the need to search for the address on the given chain.
+4. **Should we include only single-action operations, only multi-action operations, or both?**\
+We've considered including only single-action operations, where the user is only able to react with a single emoji to a sinle token, but we decided to include both single-action and multi-action operations. This way, the users can choose whether they want to emote or undo emote on a single token or on multiple tokens at once.\
```
Comment (steven2308) [inline README.md]:
```suggestion
We've considered including only single-action operations, where the user is only able to react with a single emoji to a single token, but we decided to include both single-action and multi-action operations. This way, the users can choose whether they want to emote or undo emote on a single token or on multiple tokens at once.\
```

File: README.md [hunk 2]
```diff
@@ -137,14 +299,26 @@ The impressions could have been done using user-supplied strings or numeric valu
...
+Additionally, if the collection creator wants users to record their reactions on a different chain, they can still direct the users to do just that. The repository does not validate the existence of the token being reacted to, which in theory means that you can react to non-existent token or to a token that does not exist yet. The likelyhood of a differet collection existing at the same address on another chain is significantly low, so the users can react using the collection's address on another chain and it is very unlikely that they will unintentionally react to another collection's token.
```
Comment (steven2308) [inline README.md]:
```suggestion
Additionally, if the collection creator wants users to record their reactions on a different chain, they can still direct the users to do just that. The repository does not validate the existence of the token being reacted to, which in theory means that you can react to non-existent token or to a token that does not exist yet. The likelihood of a different collection existing at the same address on another chain is significantly low, so the users can react using the collection's address on another chain and it is very unlikely that they will unintentionally react to another collection's token.
```

File: README.md [hunk 3]
```diff
@@ -100,11 +114,47 @@ interface IERC6381 /*is IERC165*/ {
...
+        address[] memory emoters,
+        address[] memory collections,
+        uint256[] memory tokenIds,
+        bytes4[] memory emojis
+    ) external view returns (bool[] memory);
+
+    /**
+     * @notice Used to get the message to be signed by the `emoter` in order for the reaction to be submitted by someone
+     *  else.
+     * @param collection The addresses of the collection smart contract containing the token being emoted at
```
Comment (steven2308) [inline README.md]:
```suggestion
     * @param collection The address of the collection smart contract containing the token being emoted at
```

File: README.md [hunk 4]
```diff
@@ -115,15 +165,127 @@ interface IERC6381 /*is IERC165*/ {
...
+    ) external;
+
+    /**
+     * @notice Used to emote or undo an emote on someone else's behalf.
+     * @dev Does nothing if attempting to set a pre-existent state.
+     * @dev MUST emit the `Emoted` event is the state of the emote is changed.
+     * @dev MUST revert if the lengths of the `collections`, `tokenIds`, `emojis` and `states` arrays are not equal.
+     * @dev MUST revert if the `deadline` has passed.
+     * @dev MUST revert if the recovered address is the zero address.
+     * @param collection Addresses of the collections containing the token being emoted at
```
Comment (steven2308) [inline README.md]:
```suggestion
     * @param collection The address of the collection smart contract containing the token being emoted at
```

File: assets/contracts/EmotableRepository.sol [hunk 5]
```diff
@@ -51,6 +110,179 @@ contract EmotableRepository is IERC6381 {
...
+        bool currentVal;
+        for (uint256 i; i < collections.length; ) {
+            currentVal = _emotesUsedByEmoter[msg.sender][collections[i]][tokenIds[i]][
+                emojis[i]
+            ] == 1;
+            if (currentVal != states[i]) {
+                if (states[i]) {
+                    _emotesPerToken[collections[i]][tokenIds[i]][emojis[i]] += 1;
+                } else {
+                    _emotesPerToken[collections[i]][tokenIds[i]][emojis[i]] -= 1;
```
Comment (steven2308) [inline assets/contracts/EmotableRepository.sol:136]: Possible optimization here and everywhere when undoing emote could be to call:
 `delete _emotesPerToken[collections[i]][tokenIds[i]][emojis[i]]`
We should gas test it

File: assets/contracts/EmotableRepository.sol [hunk 6]
```diff
@@ -51,6 +110,179 @@ contract EmotableRepository is IERC6381 {
...
+        }
+
+        bytes32 digest;
+        address signer;
+        bool currentVal;
+        for (uint256 i; i < collections.length; ) {
+            if (block.timestamp > deadlines[i]){
+                revert ExpiredPresignedEmote();
+            }
+            digest = keccak256(
```
Comment (steven2308) [inline assets/contracts/EmotableRepository.sol:285]: I think we should be able to reuse the previous method here, or extract logic. WDYT?

**Review by steven2308 — COMMENTED** [review]
I'm not fully convinced to allow bulk write operations to receive multiple collections. I feel this would be so rare that most times you will just be duplicating the collection address, which will take more gas.  Not a strong opinion, just a hunch.

Besides that, this looks pretty solid to me!

**Comment by steven2308** [comment]
> * Should we add a multi-getter for messages to be signed as well? I think that we might want to do that to provide full functionality, but at the same time, the single getter is a view function that doesn't really cause any gas waste if called multiple times thinking
It took me a while to get what you meant, but I don't see the harm in adding the bulk version. I would add it, as long as it returns many messages instead of a more complex one.
> * Should we require that the user using presigned emote provide the signer of the message? If we don't, an incorrect address could be recovered and the signature would still pass. I see the possibility for abuse here as you could simply generate new accounts and sign emotes for them and then just send those signatures and farm emotes..
Wow, that is a great catch :exploding_head:  I would add it for that reason.

File: assets/contracts/EmotableRepository.sol [hunk 7]
```diff
@@ -51,6 +110,179 @@ contract EmotableRepository is IERC6381 {
...
+    function presignedEmote(
+        address collection,
+        uint256 tokenId,
+        bytes4 emoji,
+        bool state,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public {
```
Comment (steven2308) [inline assets/contracts/EmotableRepository.sol:214]:
```suggestion
    ) public view {
```

File: assets/contracts/EmotableRepository.sol (hunk 6 above)
Comment (ThunderDeliverer) [reply assets/contracts/EmotableRepository.sol:285]: This is done to shave down on gas consumption. While calling extracted logic adds little gas, this is still cheaper. We only deploy once, so we should be ok with doing it this way, right? Similar to how we handled the properties repository.. 🤔

File: assets/contracts/EmotableRepository.sol (hunk 5 above)
Comment (ThunderDeliverer) [reply assets/contracts/EmotableRepository.sol:136]: So to just remove `-= 1`?

File: assets/contracts/EmotableRepository.sol (hunk 6 above)
Comment (steven2308) [reply assets/contracts/EmotableRepository.sol:285]: Fair enough!

File: assets/contracts/EmotableRepository.sol (hunk 5 above)
Comment (steven2308) [reply assets/contracts/EmotableRepository.sol:136]: Sorry, I had missed the `delete` keyword :see_no_evil:

**Review by steven2308 — APPROVED** [review]

**Review by steven2308 — APPROVED** [review]
Amazing!

File: assets/contracts/EmotableRepository.sol (hunk 5 above)
Comment (ThunderDeliverer) [reply assets/contracts/EmotableRepository.sol:136]: This would delete other people's reactions as well, we don't want to do that 🤔

**Review by steven2308 — APPROVED** [review]
LFG

**Pull request merged**